drivers/usb/udc/udc_bflb_bl61x.c

```diff
@@ -0,0 +1,1480 @@
+	if (cfg->speed_idx < UDC_BUS_SPEED_HS) {
+		tmp |= USB_FORCE_FS;
+	} else {
+		tmp &= ~USB_FORCE_FS;
+	}
+	sys_write32(tmp, cfg->base + USB_DEV_CTL_OFFSET);
+
+	tmp = sys_read32(cfg->base + USB_DEV_CTL_OFFSET);
+	tmp |= USB_SFRST_HOV;
+	sys_write32(tmp, cfg->base + USB_DEV_CTL_OFFSET);
+
+	/* wait for soft reset */
+	while ((sys_read32(cfg->base + USB_DEV_CTL_OFFSET)
+		& USB_SFRST_HOV) != 0) {
+	}
+
+	tmp = sys_read32(cfg->base + USB_DEV_ADR_OFFSET);
+	tmp &= ~USB_AFT_CONF;
+	sys_write32(tmp, cfg->base + USB_DEV_ADR_OFFSET);
+
+	tmp = sys_read32(cfg->base + USB_DEV_SMT_OFFSET);
+	tmp &= ~USB_SOFMT_MASK;
+	if (cfg->speed_idx == UDC_BUS_SPEED_HS) {
+		tmp |= USB_BL61X_TIMER_AFTER_RESET_HS;
+	} else {
+		tmp |= USB_BL61X_TIMER_AFTER_RESET_FS;
+	}
+	sys_write32(tmp, cfg->base + USB_DEV_SMT_OFFSET);
+
+	/* 'MISGx': Mask Interrupts Source Group x
+	 * 'ISGx' : Interrupts Source Group x (set is clear, read is status)
+	 */
+
+	/* clear irqs group 0 */
+	sys_write32(0xFFFFFFFFU, cfg->base + USB_DEV_ISG0_OFFSET);
+	/* clear irqs group 1 */
+	sys_write32(0xFFFFFFFFU, cfg->base + USB_DEV_ISG1_OFFSET);
+	/* clear irqs group 2 */
+	sys_write32(0x3FFU, cfg->base + USB_DEV_ISG2_OFFSET);
+	/* clear irqs group 3 */
+	sys_write32(0xFFFFFFFFU, cfg->base + USB_DEV_ISG3_OFFSET);
+
+	/* enable IRQs in group 0 for setup */
+	tmp = sys_read32(cfg->base + USB_DEV_MISG0_OFFSET);
+	tmp &= ~(USB_MCX_SETUP_INT);
+	tmp |= USB_MCX_COMFAIL_INT
+		| USB_MCX_COMABORT_INT
+		| USB_MCX_COMEND_INT
+		| USB_MCX_IN_INT
+		| USB_MCX_OUT_INT;
+	sys_write32(tmp, cfg->base + USB_DEV_MISG0_OFFSET);
+
+	/* disable IRQs in group 1 (fifo interrupts) */
+	sys_write32(0xFFFFFFFFU, cfg->base + USB_DEV_MISG1_OFFSET);
+
+	/* enable some group 2 interrupts */
+	sys_write32(0xFFFFFFE0U, cfg->base + USB_DEV_MISG2_OFFSET);
+
+	/* enable some group 3 interrupts (DMA completion interrupts) */
+	sys_write32(0xFFFFFFE0U, cfg->base + USB_DEV_MISG3_OFFSET);
+
+	/* enable group irqs */
+	tmp = sys_read32(cfg->base + USB_DEV_MIGR_OFFSET);
+	tmp &= ~(USB_MINT_G0
+		| USB_MINT_G1
+		| USB_MINT_G2
+		| USB_MINT_G3
+		| USB_MINT_G4);
+	sys_write32(tmp, cfg->base + USB_DEV_MIGR_OFFSET);
+
+	tmp = sys_read32(cfg->base + USB_GLB_INT_OFFSET);
+	tmp |= USB_MHC_INT;
+	tmp |= USB_MOTG_INT;
+	tmp &= ~USB_MDEV_INT;
+	sys_write32(tmp, cfg->base + USB_GLB_INT_OFFSET);
+
+	sys_write32(0xFFFFFFFFU, cfg->base + USB_DEV_EPMAP0_OFFSET);
+	sys_write32(0xFFU, cfg->base + USB_DEV_EPMAP0_OFFSET);
+	udc_bflb_bl61x_fifo_setep(dev, USB_BL61X_FIFO_EP_NONE, 1, USB_BL61X_FIFO_DIR_OUT);
+	udc_bflb_bl61x_fifo_setep(dev, USB_BL61X_FIFO_EP_NONE, 2, USB_BL61X_FIFO_DIR_OUT);
+	udc_bflb_bl61x_fifo_setep(dev, USB_BL61X_FIFO_EP_NONE, 3, USB_BL61X_FIFO_DIR_OUT);
+	udc_bflb_bl61x_fifo_setep(dev, USB_BL61X_FIFO_EP_NONE, 4, USB_BL61X_FIFO_DIR_OUT);
+
+	udc_bflb_bl61x_fifo_reset(dev, 1);
+	udc_bflb_bl61x_fifo_reset(dev, 2);
+	udc_bflb_bl61x_fifo_reset(dev, 3);
+	udc_bflb_bl61x_fifo_reset(dev, 4);
+
+	/* enable 'vdma' (virtual dma? video dma?) */
```

Comment:
BL808 [says](https://github.com/bouffalolab/bl_iot_sdk/blob/edcf613f073d3ef7d2f290d7c285b69be35da75b/components/platform/soc/bl808/bl808_e907_std/bl808_bsp_driver/std_drv/inc/bl808_usb.h#L222-L229) it's `virtual`

```c
/**
 *  @brief USB Virtual DMA transfer type definition
 */
typedef enum {
    USB_VDMA_TRANS_DIR_FIFO_2_MEM = 0,  /*!< USB VDMA transfer type is FIFO to memory */
    USB_VDMA_TRANS_DIR_MEM_2_FIFO = 1,  /*!< USB VDMA transfer type is memory to FIFO */
    USB_VDMA_TRANS_DIR_FIFO_2_FIFO = 2, /*!< USB VDMA transfer type is FIFO to FIFO */
} USB_VDMA_TRANS_DIR_Type;
```

Reply:
done